File: 907-koko-eating-bananas/koko-eating-bananas.py
class Solution(object):
    def minEatingSpeed(self, piles, h):
        l=1
        r=max(piles)
        ans=r
        while(l<=r):
            k=l+(r-l)//2
            t=0
            for i in piles:
                t+=(i+k-1)//k
                if(t>h):
                    break
            if(t<=h):
                ans=k
                r=k-1
            else:
                l=k+1
        return ans
             

File: 907-koko-eating-bananas/koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int l=1;
        int r=piles[0];
        int min =0;
        for(int i=0;i<piles.length;i++){
            if(r<piles[i])
                r=piles[i];
        }
        while(l<=r){
            int mid=l+(r-l)/2;
             int sum=0;
            for(int j=0;j<piles.length;j++){
                sum+=(piles[j]+mid-1)/mid;
                if(sum>h) break;
            }
            if(sum<=h){
                min=mid;
                r=mid-1;
            }
            else l=mid+1;
        }
        return min;
    }
}